miasm2/ir/ssa.py

```diff
@@ -339,3 +339,267 @@ def transform(self, path):
+        self._phinodes = dict()
+
+        # IRA control flow graph
+        self.graph = ira.graph
+
+    def transform(self, head):
+        """Transforms into SSA"""
+        self._init_variable_defs(head)
+        self._place_phi(head)
+        self._rename(head)
+        self._insert_phi()
+        self._convert_phi()
+
+    def reset(self):
+        """Resets SSA transformation"""
+        super(SSADiGraph, self).reset()
+        self.defs = {}
+        self._phinodes = dict()
+
+    def _init_variable_defs(self, head):
+        """
+        Initialises all variable definitions and
+        assigns the corresponding IRA blocks.
+
+        All variable definitions in self.defs contain
+        a set of IRA blocks in which the variable gets assigned
+        """
+        # architecture's instruction pointer
+        instruction_pointer = set(self.ira.arch.pc.values() + [self.ira.IRDst])
+
+        for block in self.graph.walk_depth_first_forward(head):
+            ib = self.get_block(block)
+            # blocks IR expressions
+            ir_expressions = (m2_expr.ExprAff(dst, assignblk[dst])
+                              for assignblk in ib.irs for dst in assignblk)
+            for e in ir_expressions:
+                # enforce ExprId
+                if isinstance(e.dst, m2_expr.ExprId):
+                    # exclude architecture's instruction pointer
+                    if e.dst in instruction_pointer:
+                        continue
+                    if e.dst not in self.defs:
+                        self.defs[e.dst] = set()
+                    self.defs[e.dst].add(ib.label)
+
+    def _place_phi(self, head):
+        """
+
+        For all blocks, empty phi functions will be placed for every
+        variable in the block's dominance frontier.
+
+        self.phinodes contains a dict for every block in the
+        dominance frontier. In this dict, each variable
+        definition maps to its corresponding phi function.
+
+        Source: Cytron, Ron, et al.
+        "An efficient method of computing static single assignment form"
+        Proceedings of the 16th ACM SIGPLAN-SIGACT symposium on
+        Principles of programming languages (1989), p. 30
+        """
+        # dominance frontier
+        frontier = self.graph.compute_dominance_frontier(head)
+
+        for variable in self.defs:
+            done = set()
+            todo = set()
+            intodo = set()
+
+            for block in self.defs[variable]:
+                todo.add(block)
+                intodo.add(block)
+
+            while todo:
+                block = todo.pop()
+                if block not in frontier:
+                    continue
+
+                # walk through dominance frontier of bloc
+                for node in frontier[block]:
+                    if node in done:
+                        continue
+
+                    # remember blocks which contain phi nodes
+                    if node not in self._phinodes:
```

Comment:
You can also use `setdefault` here (which does modify the value if the key already exists)